aten/src/ATen/native/TypeProperties.cpp

```diff
@@ -44,4 +44,63 @@ Tensor type_as(const Tensor& self, const Tensor& other) {
   return self.to(other.options());
 }
 
+static inline ScalarType promote_skip_undefined(ScalarType a, ScalarType b) {
+  if (a == ScalarType::Undefined) {
+    return b;
+  }
+  if (b == ScalarType::Undefined) {
+    return a;
+  }
+  return promoteTypes(a, b);
+}
+
+
+static inline ScalarType combine_categories(ScalarType higher, ScalarType lower) {
+  if (isFloatingType(higher)) {
+    return higher;
+  }
+  if (higher == ScalarType::Bool || isFloatingType(lower)) {
+    return promote_skip_undefined(higher, lower);
+  }
+  if (higher != ScalarType::Undefined) {
+      return higher;
+  }
+  return lower;
+}
+
+ScalarType _result_type(TensorList tensors) {
```

Comment:
why is this declared as a native_function if it's just a helper?

Reply:
I want to call this from TensorIterator. If it's only in the .cpp file it isn't accessible except within this file. I believe that's because native_functions causes .h files to be generated.

Reply:
just declare it in a .h file and include that .h file from TensorIterator.